### PR TITLE
libbpf-tools: Add biotop tool.

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi.

I converted `biotop` from BCC to CO-RE.
I tried to stick with the BCC behavior.

I just have two questions:
1. I only probed on some functions, is this OK? Because I think we cannot try to probe like it it is done here:
https://github.com/iovisor/bcc/blob/ca8240205e175e94cef3c70c3b5c5accbe51512f/tools/biotop.py#L178
2. I am not really satisfied with the way I handle disk name. Sadly it is a bit complicated to use a hash map in C. So if someone sees a way to improve this, feel free to share :D.

Best regards.

### CODE REVIEW

File: libbpf-tools/biotop.c
Comment (chenhengqi) [inline libbpf-tools/biotop.c:156]: Drop the surrounding `{}`.

**Comment by chenhengqi** [comment]
> Hi.
> 
> I converted `biotop` from BCC to CO-RE. I tried to stick with the BCC behavior.
> 
> I just have two questions:
> 
> 1. I only probed on some functions, is this OK? Because I think we cannot try to probe like it it is done here:
>    https://github.com/iovisor/bcc/blob/ca8240205e175e94cef3c70c3b5c5accbe51512f/tools/biotop.py#L178

Why not ? You just need to check which symbol exists and do bpf_program__set_attach_target.

> 2. I am not really satisfied with the way I handle disk name. Sadly it is a bit complicated to use a hash map in C. So if someone sees a way to improve this, feel free to share :D.
> 

This reminds me some discussions about introduce Rust-based libbpf-tools to BCC. @yonghong-song  WDYT ?

> Best regards.

File: libbpf-tools/biotop.bpf.c
Comment (mfrw) [inline libbpf-tools/biotop.bpf.c:31]: In the current kernel , the function(s) are inline [0] and instead `__blk_account_io_start` seems to be a good choice here. Do we still want to add this instead of `__blk_account_io_start` ?

[0]: [inline hot paths of blk_account_io_*()](https://lore.kernel.org/all/b0662a636bd4cc7b4f84c9d0a41efa46a688ef13.1633781740.git.asml.silence@gmail.com/)

`be6bfe36db17 block: inline hot paths of blk_account_io_*()`

File: libbpf-tools/biotop.bpf.c
Comment (mfrw) [inline libbpf-tools/biotop.bpf.c:58]: Same as above :)

**Review by mfrw — COMMENTED** [review]
A few clarification requests :)

**Comment by eiffel-fl** [comment]
Hi.

Thank you for your reviews:

>Why not ? You just need to check which symbol exists and do bpf_program__set_attach_target.

I was not aware of this function, I will then add then add all the functions and probe some of them only if symbol exists.

> This reminds me some discussions about introduce Rust-based libbpf-tools to BCC. @yonghong-song WDYT ?

I am not familiar to Rust, thus converting the C code to Rust would take me a bit of time.
But I would be happy to learn it for this contribution.

Best regards.

File: libbpf-tools/biotop.bpf.c
Comment (eiffel-fl) [reply libbpf-tools/biotop.bpf.c:31]: Hi.

I saw some of these functions were inlined into the kernel but I was not aware of how to probe only some functions (see my comment below).
I will add them and try to probe them only if their symbols exists (to filter out inlined functions).


Best regards.

Comment (chenhengqi) [reply libbpf-tools/biotop.bpf.c:31]: You can use:

https://github.com/iovisor/bcc/blob/8f40d6f57a8d94e7aee74ce358572d34d31b4ed4/libbpf-tools/trace_helpers.h#L20

File: libbpf-tools/biotop.c
Comment (eiffel-fl) [inline libbpf-tools/biotop.c:265]: Fix this.

File: libbpf-tools/biotop.c
Comment (mauriciovasquezbernal) [inline libbpf-tools/biotop.c:187]: nit: you can double the size (instead of increasing by 1) avoid doing a realloc() for each iteration of the loop.

Comment (mauriciovasquezbernal) [reply libbpf-tools/biotop.c:187]: Missing check that disk is not NULL after realloc(). (nit: you'll need to use a temp variable to avoid leaking disks is the reallocation fails).

File: libbpf-tools/biotop.c
Comment (mauriciovasquezbernal) [inline libbpf-tools/biotop.c:175]: nit: `if (!fp)`

File: libbpf-tools/biotop.c
Comment (mauriciovasquezbernal) [inline libbpf-tools/biotop.c:205]: nit: add space after `for` -> `for (size...`

File: libbpf-tools/biotop.c
Comment (mauriciovasquezbernal) [inline libbpf-tools/biotop.c:189]: nit: missing check that malloc() doesnt' return NULL. Same in other places.

File: libbpf-tools/biotop.c
Comment (mauriciovasquezbernal) [inline libbpf-tools/biotop.c:201]: I think it'll be simpler if you use a linked list for `disks`, then you won't have to realloc() the array containing pointers and so on.

**Comment by eiffel-fl** [comment]
This comments are targeting coding style.
Is there a rule regarding coding style for `*.c` files (I know `*.bpf.c` files should follow that of the kernel)?

File: libbpf-tools/biotop.c
Comment (eiffel-fl) [reply libbpf-tools/biotop.c:201]: Using a linked list would increase the time of looping through because I will not get access from the cache.
Anyway, I stated [I am not satisfied with this solution](https://github.com/iovisor/bcc/pull/3893#issuecomment-1059718669), so I would like to wait if we decide to switch from C to Rust for this control plane.

**Comment by eiffel-fl** [comment]
I am still not happy with the proposed solution but I addressed the above comments.
I think, if bcc maintainers are OK with this, we can merge this version with a C control plane and I will come back later (in another PR) with a Rust one.

File: libbpf-tools/biotop.bpf.c
Comment (mauriciovasquezbernal) [inline libbpf-tools/biotop.bpf.c:83]: When does this case happen? Should this program continue the processing even if we don't have the request on `whobyreq`?

File: libbpf-tools/biotop.bpf.c
Comment (mauriciovasquezbernal) [inline libbpf-tools/biotop.bpf.c:86]: I think there is a missing lookup, otherwise `valp` will be NULL at L95 and the counters for this iteration won't be added. 

I see there is a `bpf_map_lookup_or_try_init()` helper defined in `maps.bpf.h`, it'll simplify this code.

**Review by mauriciovasquezbernal — COMMENTED** [review]
Two comments regarding ebpf code.

File: libbpf-tools/biotop.bpf.c
Comment (eiffel-fl) [reply libbpf-tools/biotop.bpf.c:86]: Thank you for it, the code is clearer now!

File: libbpf-tools/biotop.bpf.c
Comment (eiffel-fl) [reply libbpf-tools/biotop.bpf.c:83]: I am not really sure if this can happen.
Nonetheless, this check was already present in [`biotop.py`](https://github.com/iovisor/bcc/blob/15ed997b82fa452c0ddd243fb26dff4e6e9c222f/tools/biotop.py#L171) and I tried to stick with the original code.

File: libbpf-tools/biotop.bpf.c
Comment (chenhengqi) [inline libbpf-tools/biotop.bpf.c:37]: Again, no C++ style comments.

File: libbpf-tools/biotop.bpf.c
Comment (chenhengqi) [inline libbpf-tools/biotop.bpf.c:80]: I opened a PR to handle this `request->rq_disk` properly.

https://github.com/iovisor/bcc/pull/4088/commits/0651edec65fec303438af99de96388a3e911f3ec

You can reuse it after it landed.

File: libbpf-tools/biotop.bpf.c
Comment (chenhengqi) [inline libbpf-tools/biotop.bpf.c:65]: Remove this empty line.

File: libbpf-tools/biotop.c
Comment (chenhengqi) [inline libbpf-tools/biotop.c:41]: Could this be statically allocated ? Like:
```
struct disk disks[1024];
```

So that we can avoid realloc and free below.

File: libbpf-tools/biotop.c
Comment (chenhengqi) [inline libbpf-tools/biotop.c:414]: `free(ksyms)` ?

File: libbpf-tools/biotop.c
Comment (chenhengqi) [inline libbpf-tools/biotop.c:377]: IIRC, bpf_program__set_attach_target does NOT apply to kprobe.

File: libbpf-tools/biotop.c
Comment (chenhengqi) [inline libbpf-tools/biotop.c:402]: With `bpf_program__attach_kprobe`, you don't need `biotop_bpf__attach` any more.

You should check result of `bpf_program__attach_kprobe` and free the resources on exit.

For simplicity, you can have four dummy BPF programs, so the generated skeleton will have four bpf_link s, assign the attach results to these links and call `biotop_bpf__destroy` on exit, done.

Comment (chenhengqi) [reply libbpf-tools/biotop.c:423]: @eiffel-fl  This looks incorrect to me. Please fix.

Comment (chenhengqi) [reply libbpf-tools/biotop.c:427]:
```diff
diff --git a/libbpf-tools/biotop.bpf.c b/libbpf-tools/biotop.bpf.c
index 0d418cd5..226e32d3 100644
--- a/libbpf-tools/biotop.bpf.c
+++ b/libbpf-tools/biotop.bpf.c
@@ -9,7 +9,6 @@
 #include "maps.bpf.h"
 #include "core_fixes.bpf.h"
 
-
 struct {
 	__uint(type, BPF_MAP_TYPE_HASH);
 	__uint(max_entries, 10240);
@@ -31,7 +30,7 @@ struct {
 	__type(value, struct val_t);
 } counts SEC(".maps");
 
-SEC("kprobe/blk_account_io_start")
+SEC("kprobe")
 int BPF_KPROBE(blk_account_io_start, struct request *req)
 {
 	struct who_t who = {};
@@ -57,7 +56,7 @@ int BPF_KPROBE(blk_mq_start_request, struct request *req)
 	return 0;
 }
 
-SEC("kprobe/blk_account_io_done")
+SEC("kprobe")
 int BPF_KPROBE(blk_account_io_done, struct request *req, u64 now)
 {
 	struct val_t *valp, zero = {};
diff --git a/libbpf-tools/biotop.c b/libbpf-tools/biotop.c
index 98f2fffd..2dc64658 100644
--- a/libbpf-tools/biotop.c
+++ b/libbpf-tools/biotop.c
@@ -7,6 +7,9 @@
  * Based on biotop(8) from BCC by Brendan Gregg.
  * 03-Mar-2022   Francis Laniel   Created this.
  */
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
 #include <argp.h>
 #include <errno.h>
 #include <signal.h>
@@ -70,7 +73,7 @@ void free_vector(struct vector vector) {
 	free(vector.elems);
 }
 
-struct vector disks = { 0 };
+struct vector disks = {};
 
 static volatile sig_atomic_t exiting = 0;
 
@@ -281,7 +284,7 @@ static int print_stat(struct biotop_bpf *obj)
 			printf("%8s loadavg: %s\n", ts, buf);
 		fclose(f);
 	}
-	printf("%-6s %-16s %1s %-3s %-3s %-8s %5s %7s %6s\n",
+	printf("%-7s %-16s %1s %-3s %-3s %-8s %5s %7s %6s\n",
 	       "PID", "COMM", "D", "MAJ", "MIN", "DISK", "I/O", "Kbytes", "AVGms");
 
 	while (1) {
@@ -319,7 +322,7 @@ static int print_stat(struct biotop_bpf *obj)
 		major = key->major;
 		minor = key->minor;
 
-		printf("%-6d %-16s %1s %-3d %-3d %-8s %5d %7lld %6.2f\n",
+		printf("%-7d %-16s %1s %-3d %-3d %-8s %5d %7lld %6.2f\n",
 		       key->pid, key->name, key->rwflag ? "W": "R",
 		       major, minor, search_disk_name(major, minor),
 		       value->io, value->bytes / 1024, avg_ms);
@@ -357,7 +360,6 @@ int main(int argc, char **argv)
 		.parser = parse_arg,
 		.doc = argp_program_doc,
 	};
-	struct bpf_link *probes[4] = { NULL };
 	struct biotop_bpf *obj;
 	struct ksyms *ksyms;
 	int err;
@@ -390,34 +392,22 @@ int main(int argc, char **argv)
 		goto cleanup;
 	}
 
-	if (ksyms__get_symbol(ksyms, "__blk_account_io_start")) {
-		probes[0] = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "__blk_account_io_start");
-		if (!probes[0]) {
-			warn("failed to load kprobe __blk_account_io_start\n");
-			goto cleanup;
-		}
-	}
-	if (ksyms__get_symbol(ksyms, "blk_account_io_start")) {
-		probes[1] = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "blk_account_io_start");
-		if (!probes[1]) {
-			warn("failed to load kprobe blk_account_io_start\n");
-			goto cleanup;
-		}
+	if (ksyms__get_symbol(ksyms, "__blk_account_io_start"))
+		obj->links.blk_account_io_start = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "__blk_account_io_start");
+	else
+		obj->links.blk_account_io_start = bpf_program__attach_kprobe(obj->progs.blk_account_io_start, false, "blk_account_io_start");
+	if (!obj->links.blk_account_io_start) {
+		warn("failed to attach blk_account_io_start\n");
+		goto cleanup;
 	}
 
-	if (ksyms__get_symbol(ksyms, "__blk_account_io_done")) {
-		probes[2] = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "__blk_account_io_done");
-		if (!probes[2]) {
-			warn("failed to load kprobe __blk_account_io_done\n");
-			goto cleanup;
-		}
-	}
-	if (ksyms__get_symbol(ksyms, "blk_account_io_done")) {
-		probes[3] = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "blk_account_io_done");
-		if (!probes[3]) {
-			warn("failed to load kprobe blk_account_io_done\n");
-			goto cleanup;
-		}
+	if (ksyms__get_symbol(ksyms, "__blk_account_io_done"))
+		obj->links.blk_account_io_done = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "__blk_account_io_done");
+	else
+		obj->links.blk_account_io_done = bpf_program__attach_kprobe(obj->progs.blk_account_io_done, false, "blk_account_io_done");
+	if (!obj->links.blk_account_io_done) {
+		warn("failed to attach blk_account_io_done\n");
+		goto cleanup;
 	}
 
 	err = biotop_bpf__attach(obj);
@@ -453,9 +443,6 @@ int main(int argc, char **argv)
 cleanup:
 	ksyms__free(ksyms);
 	free_vector(disks);
-	for (int i = 0; i < sizeof probes / sizeof probes[0]; i++)
-		if (probes[i])
-			bpf_link__destroy(probes[i]);
 	biotop_bpf__destroy(obj);
 
 	return err != 0;

```

Comment (chenhengqi) [reply libbpf-tools/biotop.c:427]: Try this small diff.

Comment (chenhengqi) [reply libbpf-tools/biotop.c:415]: Please paste the error here. By making `SEC("kprobe")`, we tell libbpf not to perform auto-attach, it should work with the latest libbpf submodule. Otherwise, calling `biotop_bpf__attach` will make them attach twice.

**Review by chenhengqi — APPROVED** [review]
LGTM.

**Pull request merged**
